File: cpog/pkc/pkc.py
#!/usr/bin/python3

import sys
import getopt
import datetime

import readwrite
import cnf
import ddnnf
import pog
import project

        
def usage(name):
    print("Usage: %s [-h] [-k] [-v VLEVEL] [-i FILE.cnf] [-p FILE.cpog]")
    print(" -h           Print this message")
    print(" -k           Keep intermediate files")
    print(" -v VLEVEL    Set verbosity level (0-3)")
    print(" -i FILE.cnf  Input CNF")
    print(" -p FILE.cpog Output CPOG")

def run(name, args):
    verbLevel = 1
    d4 = True
    cnfName = None
    pogName = None
    keep = False

    optlist, args = getopt.getopt(args, 'hkdv:i:n:p:')
    for (opt, val) in optlist:
        if opt == '-h':
            usage(name)
            return
        elif opt == '-k':
            keep = True
        elif opt == '-v':
            verbLevel = int(val)
        elif opt == '-d':
            d4 = True
        elif opt == '-i':
            cnfName = val
        elif opt == '-p':
            pogName = val
        else:
            print("Invalid option '%s'" % (opt))
            return

    if cnfName is None:
        print("Must give name of CNF file")
        return
    start = datetime.datetime.now()
    creader = readwrite.CnfReader(cnfName, verbLevel, False)
    pcnf = cnf.PackedCnf(creader.nvar)
    pcnf.load(creader.clauses)
    cnf.setRoot(cnfName)
    pog = project.pcnf2ppog(pcnf, creader.showVariables, verbLevel)
    if pogName is not None:
        pog.write(pogName)
        print("GEN: POG file %s written" % pogName)
    delta = datetime.datetime.now() - start
    seconds = delta.seconds + 1e-6 * delta.microseconds
    rtime = seconds - (cnf.satTime + cnf.d4Time)
    print("GEN: Elapsed time for generation: %.2f seconds" % seconds)
    print("GEN:    Time for SAT:  %.2f seconds" %  cnf.satTime)
    print("GEN:    Time for D4:   %.2f seconds" %  cnf.d4Time)
    print("GEN:    Time for rest: %.2f seconds" %  rtime)
    print("GEN %s" % project.pcache.stats())
    print("GEN: SAT solver calls = %d   D4 calls = %d" % (cnf.satCalls, cnf.d4Calls))
    if not keep:
        cnf.removeFiles()

if __name__ == "__main__":
    run(sys.argv[0], sys.argv[1:])

File: cpog/pkc/pkc.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>


#include "report.h"
#include "counters.h"
#include "project.hh"


void usage(const char *name) {
    lprintf("Usage: %s [-h] [-k] [-v VLEVEL] FORMULA.cnf FORMULA.pog\n", name);
    lprintf("  -h          Print this information\n");
    lprintf("  -k          Keep intermdiate files\n");
    lprintf("  -v VERB     Set verbosity level\n");
}

static int run(const char *cnf_name, const char *pog_name) {
    Project proj(cnf_name);
    if (verblevel >= 5) {
	printf("Initial POG:\n");
	proj.show(stdout);
    }
    proj.projecting_compile();
    if (verblevel >= 5) {
	printf("Projected POG:\n");
	proj.show(stdout);
    }
    proj.write(pog_name);
    return 0;
}

int main(int argc, char *const argv[]) {
    FILE *cnf_file = NULL;
    verblevel = 1;
    bool keep = false;
    int c;
    double start = tod();
    while ((c = getopt(argc, argv, "hkv:")) != -1) {
	switch (c) {
	case 'h':
	    usage(argv[0]);
	    return 0;
	case 'v':
	    verblevel = atoi(optarg);
	    break;
	case 'k':
	    keep = true;
	    break;
	default:
	    lprintf("Unknown comandline option '%c'\n", c);
	    usage(argv[0]);
	    break;
	}

    }
    int argi = optind;
    if (argi >= argc) {
	lprintf("Name of input CNF file required\n");
	usage(argv[0]);
	return 1;
    }
    const char *cnf_name = argv[argi++];
    if (argi >= argc) {
	lprintf("Name of output POG file required\n");
	usage(argv[0]);
	return 1;
    }
    const char *pog_name = argv[argi++];
    if (argi < argc) {
	lprintf("Unknown argument '%s'", argv[argi]);
	usage(argv[0]);
	return 1;
    }
    int result = run(cnf_name, pog_name);
    double elapsed = tod() - start;
    
    lprintf("Total execution %.2f seconds\n", elapsed);
    if (!keep)
	fmgr.flush();
    return result;
}
